### PR TITLE
fix ConfigSys and NotificationSys initialization for NAS

### PR DESCRIPTION
**Pull request by poornas**
## Description

1. globalServerConfig and globalNotificationSys are not being initialized correctly for NAS. 
2. ListenBucketNotification needs to be turned off for NAS
## Motivation and Context
 

gateways other than NAS crash when etcd is set up. 
## Regression
 Yes 
 https://github.com/minio/minio/commit/d6af3c1237d7f5f2c27585aed0ff2f4bcd0b09e3

## How Has This Been Tested?

1. start etcd (https://github.com/minio/minio/blob/master/docs/sts/etcd.md)
2. ``` export MINIO_ETCD_ENDPOINTS=http://localhost:2379```
3. ``` ./minio gateway s3```  crashes without this fix
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by harshavardhana — DISMISSED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#c1df1479-59a6-44d3-823c-5a54ec1772cd) |
<a name="c1df1479-59a6-44d3-823c-5a54ec1772cd"></a>
### 6920-207ea12/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32124
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 468fa4f6b113:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 39 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 10 seconds
(4/13) Running aws-sdk-php tests ... done in 42 seconds
(5/13) Running aws-sdk-ruby tests ... done in 3 seconds
(6/13) Running mc tests ... done in 21 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 47 seconds
(8/13) Running minio-go tests ... done in 52 seconds
(9/13) Running minio-java tests ... done in 3 minutes and 5 seconds
(10/13) Running minio-js tests ... FAILED in 30 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-3de12b3d-496d-4325-9cca-1d44d2c76ee6, events: s3:ObjectCreated:*",
  "duration": 20196,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:478:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:367:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: cmd/gateway-main.go [hunk 1]
```diff
@@ -281,10 +281,9 @@ func StartGateway(ctx *cli.Context, gw Gateway) {
 
 	// Create new notification system.
 	globalNotificationSys = NewNotificationSys(globalServerConfig, globalEndpoints)
-	if globalEtcdClient != nil {
+	if globalEtcdClient != nil && gatewayName == "nas" {
```
Comment (krishnasrinivas) [inline cmd/gateway-main.go]: you can do:
`if globalEtcdClient != nil && newObj.IsNotificationSupported() {...}`

File: cmd/xl-v1-bucket.go [hunk 2]
```diff
@@ -294,6 +294,11 @@ func (xl xlObjects) IsNotificationSupported() bool {
 	return true
 }
 
+// IsListenBucketSupported returns whether listen bucket notification is applicable for this layer.
```
Comment (krishnasrinivas) [inline cmd/xl-v1-bucket.go:297]: xlObjects is not compatible with ObjectInterface and hence you need not add this function here

File: cmd/gateway-main.go (hunk 1 above)
Comment (poornas) [reply cmd/gateway-main.go]: yes,done

File: cmd/xl-v1-bucket.go (hunk 2 above)
Comment (poornas) [reply cmd/xl-v1-bucket.go:297]: functions like savePolicyConfig take ObjectLayer as parameter, hence the need for xlObjects to implement this function.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#80b51925-2e0a-4f16-b1aa-d2f7671a146a) |
<a name="80b51925-2e0a-4f16-b1aa-d2f7671a146a"></a>
### 6920-566931c/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31172
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a3f34f0039b8:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 36 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 3 seconds
(4/13) Running aws-sdk-php tests ... done in 41 seconds
(5/13) Running aws-sdk-ruby tests ... done in 4 seconds
(6/13) Running mc tests ... done in 32 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 42 seconds
(8/13) Running minio-go tests ... done in 21 seconds
(9/13) Running minio-java tests ... done in 2 minutes and 40 seconds
(10/13) Running minio-js tests ... FAILED in 33 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-f6f88f96-6bda-4f30-9b9d-99cd00556b78, events: s3:ObjectCreated:*",
  "duration": 20077,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:478:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:367:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Review by harshavardhana — APPROVED** [review]

File: cmd/fs-v1.go [hunk 3]
```diff
@@ -1306,6 +1306,11 @@ func (fs *FSObjects) IsNotificationSupported() bool {
 	return true
 }
 
+// IsListenBucketSupported returns whether listen bucket notification is applicable for this layer.
+func (fs *FSObjects) IsListenBucketSupported() bool {
+	return true
```
Comment (harshavardhana) [inline cmd/fs-v1.go:1311]: This still enables ListenBucket you have to override this in NAS @poornas by adding another method.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/fs-v1.go (hunk 3 above)
Comment (poornas) [reply cmd/fs-v1.go:1311]: done.

**Review by krishnasrinivas — DISMISSED** [review]

**Review by harshavardhana — DISMISSED** [review]

**Review by krishnasrinivas — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#23b79331-9b46-48e4-8f38-a6db1b0932da) |
<a name="23b79331-9b46-48e4-8f38-a6db1b0932da"></a>
### 6920-ff17eca/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30581
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4339c689e0be:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 37 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 3 seconds
(4/13) Running aws-sdk-php tests ... done in 41 seconds
(5/13) Running aws-sdk-ruby tests ... done in 5 seconds
(6/13) Running mc tests ... done in 30 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 38 seconds
(8/13) Running minio-go tests ... done in 19 seconds
(9/13) Running minio-java tests ... done in 2 minutes and 13 seconds
(10/13) Running minio-js tests ... FAILED in 32 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-2bb59fee-7eb3-45fc-9b44-240a11040a97, events: s3:ObjectCreated:*",
  "duration": 20032,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:478:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:367:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Pull request merged**